libraries/protocol/custom_authority.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2018 Abit More, and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <graphene/protocol/custom_authority.hpp>
+#include <graphene/protocol/operations.hpp>
+#include <graphene/protocol/restriction_predicate.hpp>
+
+namespace graphene { namespace protocol {
+
+share_type custom_authority_create_operation::calculate_fee(const fee_parameters_type& k)const {
+   share_type core_fee_required = k.basic_fee;
+   core_fee_required += k.price_per_byte * (fc::raw::pack_size(restrictions) + fc::raw::pack_size(auth));
+   return core_fee_required;
+}
```

Comment:
I would like to avoid it. As discussed, the chain time cannot be used in the fee calculation, so both the valid-from and -to dates must always be specified, which solves the technical problem, but creates other problems. Users will be incentivized to have the validity duration be as short as possible, but it might be way in the future. So the behavior we would be incentivizing is to attempt to predict, as close to a single block interval as possible, the exact windows during which a custom authority will be used, and then publish the maximum allowable number of custom auths at a time, valid for a short window, with long intervals in between. This wastes RAM and CPU, bloats the chain history, annoys the users, and makes it so that we collect way, way less in fees for a fairly advanced and premium operation.

Users will be far happier, and fees collected far greater, if we just stick with a flat op fee and a price-per-byte.

Reply:
I would like to invite @pmconrad and @sschiessl-bcp into this discussion. I do agree that a simple per-kb fee is easier to implement and use, but it doesn't justify the costs that the chain will spend in the future due to that operation.

Reply:
From a technical perspective, did HTLC implementation have the same problems? Or what is the difference?

Reply:
Would it mitigate your concerns if we don't have an increase in fee and default of a one year time period? Fee increase would only be present for anything longer than a year (committee parameter).

The goal is to have deterministic garbage collection of abandoned accounts.

Reply:
I've created a max 1-year (committee voted) custom authority lifetime already. I forgot to put a GC in the maintenance interval for them though, I can add that.

Reply:
Ok, so we have one parameter for the maximum lifetime.

If for whatever reason, this makes RAM explode, what happens if committee reduces this lifetime to let's say a month? This would not affect already existing custom permissions, correct?

Would it make sense to introduce a duration dependent fee instead of capping the lifetime, if user wants longer than the set parameter?

Reply:
> If for whatever reason, this makes RAM explode, what happens if committee reduces this lifetime to let's say a month? This would not affect already existing custom permissions, correct?

Correct, the new limit would only come into play if the grandfathered permissions were updated.

I would only support a duration-dependent fee if it's applied at a very loose granularity, i.e. the fee bumps with every year the permission remains valid. But I'd still rather avoid it entirely, as it's inelegant to implement within our architecture (chain time can't be used to calculate the fee), and we don't do it for other database objects.

I find it improbable that this feature will be used enough to explode RAM, but even if it is, I suggest we cross that bridge when we have real data rather than hypotheticals.

Reply:
It's not the increased RAM that matters, it's the increased complexity which leads to high CPU load on every operation **times** long period that matters. That could largely impact TPS, lead to long replay time, etc.

We didn't do it for other db objects because other db objects didn't cause that much additional computation for every operation like this will potentially do.

Perhaps right now it's not the best time to think about this, but sooner or later we need to start doing similar things for other objects, for example, limit orders should have limited life time.

Reply:
So by restrict, you mean limit but not charge fees for? That's straightforward enough. So let's pick numbers...

Max number of auths for a given user and operation: 3?
Max total number of nested `restriction` elements in a custom auth: 20? 50?

I'll change the default lifetime limit to 1 month.

Reply:
> Max number of auths for a given user and operation: 3?

Sounds good.

> Max total number of nested restriction elements in a custom auth: 

Hm. How about 5?

Typical use-cases:
* price feed - no restriction required
* trade bot - perhaps a limit on markets (`amount_to_sell.asset_id` and `min_to_receive.asset_id` = 2 restrictions for a specific market), with three auths for `limit_order_create` and 2 markets per auth that would allow 6 explicitly listed markets. More if only one side (or neither) is specified.

Reply:
Can we design them as committee parameters so no need a new hard fork in case when need to change any of them?

Reply:
> Can we design them as committee parameters so no need a new hard fork in case when need to change any of them?

Yep, that's best plan.

> Hm. How about 5?

5 seems restrictively low to me... If we want it super low, 10 seems better. Take the trade bot use case: restricting the two asset IDs is already 4 restrictions:
 - Attribute assert on `amount_to_sell`
   - `asset_id` == x
 - Attribute assert on `min_to_receive`
   - `asset_id` == y

But that's not even adequate, because you'd presumably want to allow the bot to place orders on both sides of the market, so you need logical ORs:
 - Attribute assert on `amount_to_sell`
   - Logical OR:
     - Branch 1: `asset_id` == x
     - Branch 2: `asset_id` == y
 - Attribute assert on `min_to_receive`
   - Logical OR:
     - Branch 1: `asset_id` == x
     - Branch 2: `asset_id` == y

And that's already 8 restrictions. 10 seems too low to me as well, but at least it's enough room to do something nontrivial.

Reply:
> two asset IDs is already 4 restrictions

Oh, I didn't realize you counted them like that. I thought that were only 2.

> place orders on both sides of the market, so you need logical ORs

`predicate_in` could replace the ORs, which I'd count as 1.

Reply:
> `predicate_in` could replace the ORs, which I'd count as 1.

Ahh, that is true, and way more efficient. So then it stays 4 regardless of how many markets you do. Anyways, I'll set the limit to 10 for now, and make them committee-configurable.